### ISSUE TITLE
MM failures on QCFractal

### ISSUE THREAD

**Issue by dotsdl** [openforcefield/qca-dataset-submission#130]
We have recently submitted #119, but we are seeing new error modes given that this is our first MM dataset. We will document error modes observed here for triage.

**Comment by dotsdl** [comment]
New issue possibly involving charging issues with the `openforcefield` toolkit?

```
File "/data/homezvol0/tgokey/.local/miniconda3/envs/qcf/lib/python3.6/site-packages/openforcefield/typing/engines/smirnoff/parameters.py", line 3296, in postprocess_system
        raise UnassignedMoleculeChargeException(msg)
    openforcefield.typing.engines.smirnoff.parameters.UnassignedMoleculeChargeException: The following molecules did not have charges assigned by any ParameterHandler in the ForceField:
    [H][C](=[O])[N]([H])[c]1[c]([H])[c]([H])[c]([H])[c]([C]2=[C]([Br])[C]([O][C]([H])([H])[H])=[C]([H])[S]2)[c]1[H]
```

**Comment by dotsdl** [comment]
See [error cycling](https://github.com/openforcefield/qca-dataset-submission/pull/119#issuecomment-673689344) tracebacks for the kinds of failures we are currently observing.

**Comment by dotsdl** [comment]
This is my proposed plan of attack for the MM calculation failures on QCFractal we are currently observing. Please offer suggestions if I've missed any observations or if there are alternative measures we can pursue.

## What we are observing

We are seeing [error tracebacks](https://github.com/openforcefield/qca-dataset-submission/pull/119#issuecomment-673689344) like the following:

1. A small number of timeouts for getting `psi4 --version` on Lilac.

2. Many charging failure errors, perhaps universally:
    
    ```
    raise UnassignedMoleculeChargeException(msg)
    openforcefield.typing.engines.smirnoff.parameters.UnassignedMoleculeChargeException: The following molecules did not have charges assigned by any ParameterHandler in the ForceField:
    [H][c]1[c]([H])[c]([H])[c]([N]([H])[C]2([H])[C]([H])([H])[C]([H])([H])[O][C]([H])([H])[C]2([H])[H])[c]([H])[c]1[H]
    ```

## Measures to address

1. This may be due to Lilac's own cluster architecture and current usage load.
   I have asked @jchodera for options.

2. We should try and reproduce some of these locally.
   Stand up local QCArchive instance, use QCSubmit to submit the dataset to it, minus the 'default' spec.
   Observe if similar errors are produced.
   I will pursue this.

3. Continue to operate managers in production for the time being so new issues, if they exist, surface.

**Comment by j-wags** [comment]
Are the workers pulling down OpenEye? I'd encountered issues with the 2020 OE toolkit on UCI's HPC. It required a newer GLIBC than is available. I could fix it by downgrading to the `openeye-toolkits=2019`.

**Comment by j-wags** [comment]
Just checked the [worker yaml](https://anaconda.org/openforcefield/qcarchive-worker-openff/files). The above isn't the case. So it's likely an ambertools error.

**Comment by j-wags** [comment]
Next guess would be that it's somehow pulling down an old version of ambertools (pre-antechamber API/behavior change requiring the `-dr no` flag). 

Also I'm not sure about the effects of putting `defaults` at higher priority than `conda-forge`. I'll also try reproducing locally.

**Comment by trevorgokey** [comment]
The current envs are pulling 0.7.1 and amber-17. Not sure what the ramifications are to that.

**Comment by j-wags** [comment]
Ah, 17 is too old. That's very likely the issue.

**Comment by trevorgokey** [comment]
What about 17, or in other words what is changed in newer versions, that would be causing this issue?

**Comment by jthorton** [comment]
I did try and run some of the MM scans locally before submitting and they worked but I was using openeye, I think the QCEninge test we made uses ambertools and works fine so this is very strange.

**Comment by j-wags** [comment]
We should be pulling AmberTools 20.X. Antechamber enabled an input checker (acdoctor) somewhere between AT18 and 20 that chokes on carboxylates unless we specify `-dr n` in the run command. However, the `-dr` flag didn't exist before, so earlier versions crash given this input.

More details: https://github.com/openforcefield/openforcefield/issues/532

**Comment by trevorgokey** [comment]
In order to debug these, we will need to create local branches and hack away until we can find a solution. Step number one will be submitting the dataset to a local server to create a fully reproducable case.

I am looking at the recent PR with the MM issues (#119), and the provenance of how `dataset.json` was created, but the provenance of how it was actually submitted is not clear. Is there a record of how this was actually submitted?

Hopefully just an update to AT20 will fix this issue, but there are likely points that will need more debug output so we can trace the issues without have to rerun the calculation locally.

**Comment by dotsdl** [comment]
I did a sanity check and installed our [openforcefield test env](https://github.com/openforcefield/openforcefield/blob/master/devtools/conda-envs/test_env.yaml). I get `ambertools` 0.17.0 in that too, sourced from `pypi` it seems:
```
# packages in environment at /home/david/.conda/envs/openff-test-env:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                 conda_forge    conda-forge
_openmp_mutex             4.5                       1_gnu    conda-forge
abseil-cpp                20200225.2           he1b5a44_2    conda-forge
alabaster                 0.7.12                     py_0    conda-forge
amberlite                 16.0                     pypi_0    pypi
ambertools                17.0                     pypi_0    pypi
argon2-cffi               20.1.0           py37h8f50634_1    conda-forge
arpack                    3.7.0                hc6cf775_1    conda-forge
arrow-cpp                 1.0.0           py37h1234567_1_cpu    conda-forge
astor                     0.8.1              pyh9f0ad1d_0    conda-forge
attrs                     19.3.0                     py_0    conda-forge
aws-sdk-cpp               1.7.164              hba45d7a_2    conda-forge
babel                     2.8.0                      py_0    conda-forge
backcall                  0.2.0              pyh9f0ad1d_0    conda-forge
backports                 1.0                        py_2    conda-forge
backports.functools_lru_cache 1.6.1                      py_0    conda-forge
bleach                    3.1.5              pyh9f0ad1d_0    conda-forge
blosc                     1.20.0               he1b5a44_0    conda-forge
boost                     1.72.0           py37h9de70de_0    conda-forge
boost-cpp                 1.72.0               h7b93d67_2    conda-forge
brotli                    1.0.7             he1b5a44_1004    conda-forge
brotlipy                  0.7.0           py37h8f50634_1000    conda-forge
bson                      0.5.9                      py_0    conda-forge
bzip2                     1.0.8                h516909a_2    conda-forge
c-ares                    1.16.1               h516909a_0    conda-forge
ca-certificates           2020.6.20            hecda079_0    conda-forge
cairo                     1.16.0            h3fc0475_1005    conda-forge
certifi                   2020.6.20        py37hc8dfbb8_0    conda-forge
cffi                      1.14.1           py37h2b28604_0    conda-forge
cftime                    1.2.1            py37h03ebfcd_0    conda-forge
chardet                   3.0.4           py37hc8dfbb8_1006    conda-forge
codecov                   2.1.8              pyh9f0ad1d_0    conda-forge
coverage                  4.5.4            py37h516909a_0    conda-forge
cryptography              3.0              py37hb09aad4_0    conda-forge
curl                      7.71.1               he644dc0_4    conda-forge
cycler                    0.10.0                     py_2    conda-forge
cython                    0.29.21          py37h3340039_0    conda-forge
decorator                 4.4.2                      py_0    conda-forge
defusedxml                0.6.0                      py_0    conda-forge
docutils                  0.16             py37hc8dfbb8_1    conda-forge
entrypoints               0.3             py37hc8dfbb8_1001    conda-forge
fftw                      3.3.8           nompi_h7f3a6c3_1111    conda-forge
fftw3f                    3.3.4                         2    omnia
fontconfig                2.13.1            h1056068_1002    conda-forge
freetype                  2.10.2               he06d7ca_0    conda-forge
gettext                   0.19.8.1          hc5be6a0_1002    conda-forge
gflags                    2.2.2             he1b5a44_1004    conda-forge
glib                      2.65.0               h6f030ca_0    conda-forge
glog                      0.4.0                h49b9bf7_3    conda-forge
grpc-cpp                  1.30.2               heedbac9_0    conda-forge
h5py                      2.10.0          nompi_py37h90cd8ad_104    conda-forge
hdf4                      4.2.13            hf30be14_1003    conda-forge
hdf5                      1.10.6          nompi_h3c11f04_101    conda-forge
icu                       67.1                 he1b5a44_0    conda-forge
idna                      2.10               pyh9f0ad1d_0    conda-forge
imagesize                 1.2.0                      py_0    conda-forge
importlib-metadata        1.7.0            py37hc8dfbb8_0    conda-forge
importlib_metadata        1.7.0                         0    conda-forge
iniconfig                 1.0.1              pyh9f0ad1d_0    conda-forge
ipykernel                 5.3.4            py37h43977f1_0    conda-forge
ipython                   7.17.0           py37hc6149b9_0    conda-forge
ipython_genutils          0.2.0                      py_1    conda-forge
ipywidgets                7.5.1                      py_0    conda-forge
jedi                      0.17.2           py37hc8dfbb8_0    conda-forge
jinja2                    2.11.2             pyh9f0ad1d_0    conda-forge
jpeg                      9d                   h516909a_0    conda-forge
jsonschema                3.2.0            py37hc8dfbb8_1    conda-forge
jupyter_client            6.1.6                      py_0    conda-forge
jupyter_core              4.6.3            py37hc8dfbb8_1    conda-forge
kiwisolver                1.2.0            py37h99015e2_0    conda-forge
krb5                      1.17.1               hfafb76e_2    conda-forge
latexcodec                2.0.1              pyh9f0ad1d_0    conda-forge
lcms2                     2.11                 hbd6801e_0    conda-forge
ld_impl_linux-64          2.34                 hc38a660_9    conda-forge
libblas                   3.8.0               17_openblas    conda-forge
libcblas                  3.8.0               17_openblas    conda-forge
libcurl                   7.71.1               hcdd3856_4    conda-forge
libedit                   3.1.20191231         h46ee950_1    conda-forge
libev                     4.33                 h516909a_0    conda-forge
libevent                  2.1.10               hcdb4288_1    conda-forge
libffi                    3.2.1             he1b5a44_1007    conda-forge
libgcc-ng                 9.3.0               h24d8f2e_14    conda-forge
libgfortran-ng            7.5.0               hdf63c60_14    conda-forge
libgomp                   9.3.0               h24d8f2e_14    conda-forge
libiconv                  1.15              h516909a_1006    conda-forge
liblapack                 3.8.0               17_openblas    conda-forge
libnetcdf                 4.7.4           nompi_h84807e1_105    conda-forge
libnghttp2                1.41.0               hab1572f_1    conda-forge
libopenblas               0.3.10          pthreads_hb3c22a3_4    conda-forge
libpng                    1.6.37               hed695b0_2    conda-forge
libprotobuf               3.12.4               h8b12597_0    conda-forge
libsodium                 1.0.18               h516909a_0    conda-forge
libssh2                   1.9.0                hab1572f_5    conda-forge
libstdcxx-ng              9.3.0               hdf63c60_14    conda-forge
libtiff                   4.1.0                hc7e4089_6    conda-forge
libutf8proc               2.5.0                h516909a_2    conda-forge
libuuid                   2.32.1            h14c3975_1000    conda-forge
libwebp-base              1.1.0                h516909a_3    conda-forge
libxcb                    1.13              h14c3975_1002    conda-forge
libxml2                   2.9.10               h72b56ed_2    conda-forge
libxslt                   1.1.33               h572872d_1    conda-forge
lxml                      4.5.2            py37he3881c9_0    conda-forge
lz4-c                     1.9.2                he1b5a44_1    conda-forge
lzo                       2.10              h14c3975_1000    conda-forge
markupsafe                1.1.1            py37h8f50634_1    conda-forge
matplotlib-base           3.3.0            py37hd478181_1    conda-forge
mdtraj                    1.9.4            py37h4112681_0    conda-forge
mistune                   0.8.4           py37h8f50634_1001    conda-forge
mmpbsa-py                 16.0                     pypi_0    pypi
mock                      4.0.2            py37hc8dfbb8_0    conda-forge
more-itertools            8.4.0                      py_0    conda-forge
mpiplus                   v0.0.1                py37_1000    conda-forge
msgpack-python            1.0.0            py37h99015e2_1    conda-forge
nbconvert                 5.6.1            py37hc8dfbb8_1    conda-forge
nbformat                  5.0.7                      py_0    conda-forge
nbval                     0.9.6              pyh9f0ad1d_0    conda-forge
ncurses                   6.2                  he1b5a44_1    conda-forge
netcdf-fortran            4.5.3           nompi_hfef6a68_100    conda-forge
netcdf4                   1.5.4           nompi_py37hdc49583_100    conda-forge
networkx                  2.4                        py_1    conda-forge
nglview                   2.7.7              pyh5ca1d4c_1    conda-forge
notebook                  6.1.3            py37hc8dfbb8_0    conda-forge
numexpr                   2.7.1            py37h0da4684_1    conda-forge
numpy                     1.19.1           py37h8960a57_0    conda-forge
olefile                   0.46                       py_0    conda-forge
openeye-toolkits          2020.0.4                 py37_0    openeye
openforcefield            0.7.1              pyh39e3cac_0    omnia
openforcefields           1.2.0                    py37_0    omnia
openmm                    7.4.2           py37_cuda101_rc_1    omnia
openmmforcefields         0.8.0                    py37_0    omnia
openmmtools               0.20.0                   py37_0    omnia
openssl                   1.1.1g               h516909a_1    conda-forge
oset                      0.1.3                      py_1    conda-forge
packaging                 20.4               pyh9f0ad1d_0    conda-forge
packmol-memgen            1.1.0rc0                 pypi_0    pypi
pandas                    1.1.0            py37h3340039_0    conda-forge
pandoc                    2.10.1               h516909a_0    conda-forge
pandocfilters             1.4.2                      py_1    conda-forge
parmed                    at20RC5+54.g5702a232fe.dirty          pypi_0    pypi
parquet-cpp               1.5.1                         2    conda-forge
parso                     0.7.1              pyh9f0ad1d_0    conda-forge
pcre                      8.44                 he1b5a44_0    conda-forge
pdb4amber                 1.7.dev0                 pypi_0    pypi
perl                      5.26.2            h516909a_1006    conda-forge
pexpect                   4.8.0            py37hc8dfbb8_1    conda-forge
pickleshare               0.7.5           py37hc8dfbb8_1001    conda-forge
pillow                    7.2.0            py37h718be6c_1    conda-forge
pint                      0.14                       py_0    conda-forge
pip                       20.2.2                     py_0    conda-forge
pixman                    0.38.0            h516909a_1003    conda-forge
plotly                    4.9.0              pyh9f0ad1d_0    conda-forge
pluggy                    0.13.1           py37hc8dfbb8_2    conda-forge
prometheus_client         0.8.0              pyh9f0ad1d_0    conda-forge
prompt-toolkit            3.0.6                      py_0    conda-forge
psutil                    5.7.2            py37h8f50634_0    conda-forge
pthread-stubs             0.4               h14c3975_1001    conda-forge
ptyprocess                0.6.0                   py_1001    conda-forge
py                        1.9.0              pyh9f0ad1d_0    conda-forge
py-cpuinfo                7.0.0              pyh9f0ad1d_0    conda-forge
pyarrow                   1.0.0           py37h1234567_1_cpu    conda-forge
pybtex                    0.22.2           py37hc8dfbb8_1    conda-forge
pybtex-docutils           0.2.2            py37hc8dfbb8_1    conda-forge
pycairo                   1.19.1           py37h01af8b0_3    conda-forge
pycparser                 2.20               pyh9f0ad1d_2    conda-forge
pydantic                  1.6.1            py37h8f50634_0    conda-forge
pygments                  2.6.1                      py_0    conda-forge
pymbar                    3.0.5            py37hc1659b7_0    conda-forge
pyopenssl                 19.1.0                     py_1    conda-forge
pyparsing                 2.4.7              pyh9f0ad1d_0    conda-forge
pyrsistent                0.16.0           py37h8f50634_0    conda-forge
pysocks                   1.7.1            py37hc8dfbb8_1    conda-forge
pytables                  3.6.1            py37h56451d4_2    conda-forge
pytest                    6.0.1            py37hc8dfbb8_0    conda-forge
pytest-cov                2.10.0             pyh9f0ad1d_0    conda-forge
python                    3.7.8           h6f2ec95_1_cpython    conda-forge
python-dateutil           2.8.1                      py_0    conda-forge
python_abi                3.7                     1_cp37m    conda-forge
pytraj                    2.0.5                    pypi_0    pypi
pytz                      2020.1             pyh9f0ad1d_0    conda-forge
pyyaml                    5.3.1            py37h8f50634_0    conda-forge
pyzmq                     19.0.2           py37hac76be4_0    conda-forge
qcelemental               0.15.1                     py_0    conda-forge
qcengine                  0.15.0                     py_0    conda-forge
qcportal                  0.13.1                     py_0    conda-forge
rdkit                     2020.03.5        py37h3f38cc2_0    conda-forge
re2                       2020.08.01           he1b5a44_0    conda-forge
readline                  8.0                  he28a2e2_2    conda-forge
requests                  2.24.0             pyh9f0ad1d_0    conda-forge
retrying                  1.3.3                      py_2    conda-forge
sander                    16.0                     pypi_0    pypi
scipy                     1.5.2            py37hb14ef9d_0    conda-forge
send2trash                1.5.0                      py_0    conda-forge
setuptools                49.4.0           py37hc8dfbb8_0    conda-forge
six                       1.15.0             pyh9f0ad1d_0    conda-forge
smirnoff99frosst          1.1.0                    py37_1    omnia
snappy                    1.1.8                he1b5a44_3    conda-forge
snowballstemmer           2.0.0                      py_0    conda-forge
sphinx                    3.2.0                      py_0    conda-forge
sphinxcontrib-applehelp   1.0.2                      py_0    conda-forge
sphinxcontrib-bibtex      1.0.0                      py_0    conda-forge
sphinxcontrib-devhelp     1.0.2                      py_0    conda-forge
sphinxcontrib-htmlhelp    1.0.3                      py_0    conda-forge
sphinxcontrib-jsmath      1.0.1                      py_0    conda-forge
sphinxcontrib-qthelp      1.0.3                      py_0    conda-forge
sphinxcontrib-serializinghtml 1.1.4                      py_0    conda-forge
sqlite                    3.32.3               h4cf870e_1    conda-forge
terminado                 0.8.3            py37hc8dfbb8_1    conda-forge
testpath                  0.4.4                      py_0    conda-forge
thrift-cpp                0.13.0               h62aa4f2_2    conda-forge
tinydb                    4.1.1              pyh9f0ad1d_0    conda-forge
tk                        8.6.10               hed695b0_0    conda-forge
toml                      0.10.1             pyh9f0ad1d_0    conda-forge
tornado                   6.0.4            py37h8f50634_1    conda-forge
tqdm                      4.48.2             pyh9f0ad1d_0    conda-forge
traitlets                 4.3.3            py37hc8dfbb8_1    conda-forge
urllib3                   1.25.10                    py_0    conda-forge
wcwidth                   0.2.5              pyh9f0ad1d_1    conda-forge
webencodings              0.5.1                      py_1    conda-forge
wheel                     0.34.2                     py_1    conda-forge
widgetsnbextension        3.5.1            py37hc8dfbb8_1    conda-forge
xmltodict                 0.12.0                     py_0    conda-forge
xorg-kbproto              1.0.7             h14c3975_1002    conda-forge
xorg-libice               1.0.10               h516909a_0    conda-forge
xorg-libsm                1.2.3             h84519dc_1000    conda-forge
xorg-libx11               1.6.11               h516909a_0    conda-forge
xorg-libxau               1.0.9                h14c3975_0    conda-forge
xorg-libxdmcp             1.1.3                h516909a_0    conda-forge
xorg-libxext              1.3.4                h516909a_0    conda-forge
xorg-libxrender           0.9.10            h516909a_1002    conda-forge
xorg-libxt                1.2.0                h516909a_0    conda-forge
xorg-renderproto          0.11.1            h14c3975_1002    conda-forge
xorg-xextproto            7.3.0             h14c3975_1002    conda-forge
xorg-xproto               7.0.31            h14c3975_1007    conda-forge
xz                        5.2.5                h516909a_1    conda-forge
yaml                      0.2.5                h516909a_0    conda-forge
zeromq                    4.3.2                he1b5a44_3    conda-forge
zipp                      3.1.0                      py_0    conda-forge
zlib                      1.2.11            h516909a_1007    conda-forge
zstd                      1.4.5                h6597ccf_2    conda-forge
```

**Comment by dotsdl** [comment]
I'm trying to narrow down which of the dependencies in [qcarchive-worker-openff](https://anaconda.org/openforcefield/qcarchive-worker-openff/files) cause us to grab ambertools 17 over 20.x. I'll report my findings here.

**Comment by dotsdl** [comment]
Do we absolutely need Python <= 3.7? Trying to eliminate constraints where possible.

**Comment by trevorgokey** [comment]
So I just did a new env creation (`conda env create -n tmp qcarchive/qcarchive-worker-openff`), and it was pulling AT20 from conda forge. Perhaps it's a channel priority thing? This host is using conda 4.8.3

**Comment by j-wags** [comment]
> Do we absolutely need Python <= 3.7? Trying to eliminate constraints where possible.

The OpenFF toolkit will need to support py3.6 until [Feb 2022](https://github.com/openforcefield/openforcefield/blob/edbec2bd9252db269b18798c0322ec2a0a5ae12a/.github/workflows/py36-checks.yml#L19-L26). That said, these workers are free to run whatever version of python you like.

**Comment by trevorgokey** [comment]
I actually just recreated the worker env, and it is pulling AT20. So it might be the case that AT20 is failing on these molecules, but AT17 is not.

**Comment by j-wags** [comment]
There is no pip package for ambertools -- The "version 17      pypi" were seeing is due to how AmberTools, where the conda package has to install it via pip. So the version reported by `conda list` may also not be reliable.

**Comment by dotsdl** [comment]
I just ran `conda env create -n tmp qcarchive/qcarchive-worker-openff` as @trevorgokey did above, with conda 4.8.3, and I get `ambertools` 17.0.

I don't know which way is up now.

Trevor, do you have a `~/.condarc` file? I don't *think* it should impact things for an environment file, but I'm also not certain of that. I deleted mine before running this.

**Comment by j-wags** [comment]
I'm able to reproduce this specific error locally in two ways:
1) Delete AmberTools and not install OpenEye
2) `del ff._parameter_handlers['ToolkitAM1BCC']` and not providing `charge_from_molecules` to `ff.create_openmm_system`

I wasn't able to find any values I could pass to `charge_from_molecules` that would lead to the observed error. It's possible that there's something getting (exotically) mangled in this molecule pipeline, leading to an unusual mis-perception somewhere in the middle, or some very weird value getting passed to `charge_from_molecules`.

My top suspect is an ambertools deployment issue. However we all built environments from this yaml and none of us could reproduce the error, not even on UCI HPC where the known errors are being raised.

**Comment by j-wags** [comment]
Also worth noting: This deployment resolves to AmberTools 20.4, which is strange because I've been pulling 20.6 all week, and 20.7 came out yesterday. Doesn't seem like it'd be a huge issue, but this may help us pinpoint the underlying cause.

**Comment by trevorgokey** [comment]
I found that the pypi version that AT17 installs overwrites the binaries (sqm) from AT20, so at the current moment a `conda install ambertools=20 --force-reinstall` is necessary to have the newer sqm visible to antechamber.

**Comment by dotsdl** [comment]
Thank you so much for identifying this @trevorgokey! You've completely unblocked us here!

**Comment by trevorgokey** [comment]
It still doesn't make sense though, since reinstalling AT20 should just do the same thing (install its own miniconda).. I'll take another dive relatively soon.

**Comment by dotsdl** [comment]
Closing, since we have discovered the cause of this issue and have put a workaround in place. 

@trevorgokey, can you add your observations to [this ambertools conda-forge feedstock issue](https://github.com/conda-forge/ambertools-feedstock/issues/45) so that an upstream fix can be pursued?

**Comment by j-wags** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Automate FF release simulation-based "canary" tests](https://trello.com/c/s4t9DE7B/114-automate-ff-release-simulation-based-canary-tests)
